helix-term/src/ui/editor/diagnostics_annotations.rs

```diff
@@ -0,0 +1,282 @@
+        // displaying the last one and go up one at a time
+        let mut pos_y = viewport
+            .y
+            .saturating_add(pos.visual_line)
+            .saturating_add(line_count);
+
+        // Then we iterate the stack we just built to find diagnostics
+        for (idx, item) in stack.iter().enumerate() {
+            let (text, style) = match item {
+                StackItem::Diagnostic(text, style) => (text.trim(), *style),
+                _ => continue,
+            };
+
+            left.clear();
+            let mut overlap = false;
+            let mut multi = 0;
+
+            // Iterate the stack for this line to find elements on the left.
+            let mut peekable = stack[..idx].iter().peekable();
+            while let Some(item2) = peekable.next() {
+                match item2 {
+                    &StackItem::Space(n) if multi == 0 => left.push(DisplayItem::Space(n)),
+                    &StackItem::Space(n) => {
+                        left.push(DisplayItem::String("─".repeat(n as usize), style, n))
+                    }
+                    StackItem::Blank(_) => {
+                        left.push(DisplayItem::Static(
+                            if multi == 0 { "└" } else { "┴" },
+                            style,
+                            1,
+                        ));
+                        multi += 1;
+                    }
+                    StackItem::Diagnostic(_, style) => {
+                        // If an overlap follows this, don't add an extra column.
+                        if !(matches!(peekable.peek(), Some(StackItem::Overlap))) {
+                            left.push(DisplayItem::Static("│", *style, 1));
+                        }
+                        overlap = false;
+                    }
+                    StackItem::Overlap => overlap = true,
+                }
+            }
+
+            let center_symbol = if overlap && multi > 0 {
+                "┼─── "
+            } else if overlap {
+                "├─── "
+            } else if multi > 0 {
+                "┴─── "
+            } else {
+                "└─── "
+            };
+
+            center.clear();
+            center.push(DisplayItem::Static(center_symbol, style, 5));
+
+            // TODO: We can draw on the left side if and only if:
+            // a. Is the last one stacked this line.
+            // b. Has enough space on the left.
+            // c. Is just one line.
+            // d. Is not an overlap.
+
+            let lines_offset = text.lines().count() as u16;
+            pos_y -= lines_offset;
+
+            // Use `view` since it's the whole outer view instead of just the inner area so that the background
+            // is also applied to the gutters and other elements that are not in the editable part of the document
+            let diag_area = Rect::new(
+                whole_view_aera.x,
+                pos_y + 1,
```

Comment:
This can go out of bounds, aka. higher than the `viewport.height` which causes a panic.
I have debugged this as i am using this in my daily-driver fork.

a simple fix would be some kind of clamp of `pos_y` (also on L#247 ) to the `viewport.height`